src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+  (let (
+      (wrapped-name-props (map-get? name-properties ((namespace namespace) (name name))))
+      (current-owner (map-get? owner-name ((owner contract-caller))))
+      (namespace-props (unwrap! (map-get? namespaces ((namespace namespace))) (ok false))))
+    ;; Ensure that namespace has been launched 
+    (unwrap! (get launched-at namespace-props) (ok false))
+    ;; Early return - Name has never be minted
+    (asserts! (is-none (nft-get-owner? names (tuple (name name) (namespace namespace)))) (ok true))
+    ;; Integrity check - Ensure that we have some entries in nft && owner-name && name-props
+    (asserts! 
+      (and 
+        (is-some wrapped-name-props)
+        (is-some current-owner))
+      (err ERR_PANIC))
+    (let ((name-props (unwrap! wrapped-name-props (err ERR_PANIC))))
+      ;; Early return - Name has been revoked and can be registered
+      (asserts! (is-some (get revoked-at name-props)) (ok true))
+      ;; Integrity check - Ensure that the name was either "imported" or "registered".
+      (asserts! 
+        (or 
+          (and (is-some (get registered-at name-props)) (is-none (get imported-at name-props)))
+          (and (is-some (get imported-at name-props)) (is-none (get registered-at name-props))))
+        (err ERR_PANIC))
+      ;; Is lease expired?
+      (is-name-lease-expired? namespace name))))
+
+(define-read-only (name-resolve (namespace (buff 19)) (name (buff 16)))
+  (let (
+    (owner (unwrap!
+      (nft-get-owner? names (tuple (name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+    (name-props (unwrap!
+      (map-get? name-properties ((name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND)))
+    (namespace-props (unwrap! 
+      (map-get? namespaces ((namespace namespace))) 
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (is-lease-expired (is-name-lease-expired? namespace name)))
+    ;; The namespace must be launched
+    (asserts!
+      (is-some (get launched-at namespace-props))
+      (err ERR_NAMESPACE_NOT_LAUNCHED))
+    ;; The name must not be in the renewal grace period
+    (asserts!
+      (is-eq (unwrap! (is-name-in-grace-period? namespace name) (err ERR_PANIC)) false)
+      (err ERR_NAME_GRACE_PERIOD))
+    ;; The name must not be expired
+    (if (is-ok is-lease-expired)
+      (asserts! (not (unwrap! is-lease-expired (err ERR_PANIC))) (err ERR_NAME_EXPIRED))
+      true)
+    ;; The name must not be revoked
+    (asserts!
+      (is-none (get revoked-at name-props))
+      (err ERR_NAME_REVOKED))
+    ;; Get the zonefile
+    (ok (get zonefile-hash name-props))))
+
+;; (begin
+;;   (ft-mint? stx u10000000000 'SP2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKNRV9EJ7)
+;;   (ft-mint? stx u10000000 'S02J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKPVKG2CE)
+;;   (ft-mint? stx u10000000 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+;;   (ft-mint? stx u10000000 'SPMQEKN07D1VHAB8XQV835E3PTY3QWZRZ5H0DM36))
```

Comment:
Are there tests anywhere for BNS?

Reply:
Yeah that's the sad part of this PR. For the context, the BNS contract has been written a year ago. I used to have a fair amount of testing, but the tests were written with the Clarity JS SDK - which appeared to be broken.
I'll commit the tests here https://github.com/blockstack/stacks-blockchain/pull/2013/commits/09fc2b43a53e3f60f62de1a39ae80e4f46efaa45 for now, and will talk with the team and see if someone with bandwidth can help me fixing the SDK.

Reply:
I was able to recover my tests (thank you so much for your help @hstove!), they look green, I'll add more test cases, testing for the latest modifications, and push.